### PR TITLE
Returned Internal Service Error when IAM fails

### LINKED ISSUE #195: KG returns 401 Unauthorized when call to IAM fails
This is the response which is returned by KG with 401 status code when call to IAM fails: 
```
{
  @context: "https://bbp-nexus.epfl.ch/dev/v0/contexts/nexus/core/error/v0.1.0",
  code: "UnauthorizedAccess"
}
```
It should return an error message with 500 status code

### PR DESCRIPTION
**Pull request by umbreak**
Fixes #195

### CODE REVIEW

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/RejectionHandling.scala [hunk 1]
```diff
@@ -45,6 +46,9 @@ object RejectionHandling {
           complete(BadRequest -> (e: CommonRejections))
         case ValidationRejection(_, Some(e: IllegalVersionFormat)) =>
           complete(BadRequest -> (e: CommonRejections))
+        case CustomAuthorizationRejection(e) =>
+        complete(InternalServerError -> (e: CommonRejections))
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/RejectionHandling.scala]: formatting

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/AuthDirectives.scala [hunk 2]
```diff
@@ -58,10 +57,19 @@ trait AuthDirectives {
     */
   def authenticateCaller(implicit iamClient: IamClient[Future], cred: Option[OAuth2BearerToken]): Directive1[Caller] =
     onComplete(iamClient.getCaller(cred, filterGroups = true)).flatMap {
-      case Success(caller) => provide(caller)
-      case _               => reject(AuthorizationFailedRejection)
+      case Success(caller)             => provide(caller)
+      case Failure(UnauthorizedAccess) => reject(AuthorizationFailedRejection)
+      case Failure(err)                => reject(CustomAuthorizationRejection(UnderlyingServiceError(err.getMessage)))
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/AuthDirectives.scala]: err.getMessage can throw if the message is null

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/AuthDirectives.scala [hunk 3]
```diff
@@ -27,16 +31,12 @@ trait AuthDirectives {
...
-        .recoverWith {
-          case UnauthorizedAccess                      => Future.successful(false)
-          case err: UnexpectedUnsuccessfulHttpResponse => Future.failed(err)
-        }
+                                                          cred: Option[OAuth2BearerToken]): Directive0 =
+    onComplete(iamClient.getAcls(prefix ++ resource)).flatMap {
+      case Success(acls) if (acls.permissions.contains(perm)) => pass
+      case Success(_)                                         => reject(AuthorizationFailedRejection)
+      case Failure(UnauthorizedAccess)                        => reject(AuthorizationFailedRejection)
+      case Failure(err)                                       => reject(CustomAuthorizationRejection(UnderlyingServiceError(err.getMessage)))
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/AuthDirectives.scala]: err.getMessage can throw if the message is null

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/CommonRejections.scala [hunk 4]
```diff
@@ -29,4 +29,12 @@ object CommonRejections {
   @SuppressWarnings(Array("IncorrectlyNamedExceptions"))
   final case class IllegalVersionFormat(override val message: String) extends Err(message) with CommonRejections
 
+  /**
+    * Signals the inability to connect to an underlying service to perform a request
+    *
+    * @param message a human readable description of the cause
+    */
+  @SuppressWarnings(Array("IncorrectlyNamedExceptions"))
+  final case class UnderlyingServiceError(override val message: String) extends Err(message) with CommonRejections
```
Comment (wwajerowicz) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/CommonRejections.scala]: I think Downstream or Upstream would be better word than Underlying

**Pull request merged**